services/plugins/npmjs-plugins-source.ts

```diff
@@ -0,0 +1,113 @@
+import * as cheerio from "cheerio";
+import {PluginsSourceBase} from "./plugins-source-base";
+
+export class NpmjsPluginsSource extends PluginsSourceBase implements IPluginsSource {
+	private static NPMJS_ADDRESS = "http://npmjs.org";
+
+	private _hasReturnedTheInitialPlugins: boolean;
+	private _keywords: string[];
+
+	constructor(private $httpClient: Server.IHttpClient,
+		private $childProcess: IChildProcess,
+		private $hostInfo: IHostInfo,
+		private $progressIndicator: IProgressIndicator,
+		private $logger: ILogger,
+		private $errors: IErrors) {
+		super();
+	}
+
+	public initialize(projectDir: string, keywords: string[]): IFuture<void> {
+		return (() => {
+			super.initialize(projectDir, keywords).wait();
+
+			this._keywords = keywords;
+
+			this._plugins = this.getPluginsFromNpmjs(keywords, 0).wait();
+		}).future<void>()();
+	}
+
+	public getPlugins(page: number, count: number): IFuture<IBasicPluginInformation[]> {
+		return ((): IBasicPluginInformation[] => {
+			if (!this._hasReturnedTheInitialPlugins && page === 0) {
+				return this._plugins;
+			}
+
+			let result = this.getPluginsFromNpmjs(this._keywords, page).wait();
+
+			this._plugins = this._plugins.concat(result);
+
+			return result;
+		}).future<IBasicPluginInformation[]>()();
+	}
+
+	public getAllPlugins(): IFuture<IBasicPluginInformation[]> {
+		return ((): IBasicPluginInformation[] => {
+			let result: IBasicPluginInformation[] = [];
+
+			let currentPluginsFound: IBasicPluginInformation[] = [];
+			let page = 0;
+
+			do {
+				currentPluginsFound = this.getPluginsFromNpmjs(this._keywords, page++).wait();
+				if (currentPluginsFound && currentPluginsFound.length) {
+					result = result.concat(currentPluginsFound);
+				}
+			} while (currentPluginsFound && currentPluginsFound.length);
+
+			return result;
+		}).future<IBasicPluginInformation[]>()();
+	}
+
+	public hasPlugins(): boolean {
+		return !!(this._plugins && this._plugins.length);
+	}
+
+	private prepareScopedPluginName(keywords: string[]): string {
+		let pluginName = keywords[0];
+
+		pluginName = pluginName.replace("@", "%40");
+		pluginName = pluginName.replace("/", "%2F");
+
+		return pluginName;
+	}
+
+	private getPluginsFromNpmjs(keywords: string[], page: number): IFuture<IBasicPluginInformation[]> {
+		return ((): IBasicPluginInformation[] => {
+			let pluginName = this.prepareScopedPluginName(keywords);
+			let url = `${NpmjsPluginsSource.NPMJS_ADDRESS}/search?q=${pluginName}&page=${page}`;
+
+			try {
+				let responseBody: string = this.$httpClient.httpRequest(url).wait().body;
+				let $ = cheerio.load(responseBody);
```

Comment:
the `cheerio` module might be a bit overkill for what we're trying to achieve here. You can take a look at [parser5](https://github.com/inikulin/parse5) or [htmlparser2](https://github.com/fb55/htmlparser2) and see if we can't tailor that for our needs